multi-region-load-tests/multi-region-tests.py

```diff
@@ -0,0 +1,332 @@
+import io
+import zipfile
+from pathlib import Path
+import pandas as pd
+import requests
+import configparser
+import time
+import uuid
+
+from azure.identity import ClientSecretCredential
+from azure.mgmt.loadtesting import LoadTestMgmtClient
+from azure.developer.loadtesting import LoadTestRunClient
+from azure.developer.loadtesting import LoadTestAdministrationClient
+from azure.storage.blob import BlobClient
+from azure.core.exceptions import HttpResponseError
+
+config = configparser.ConfigParser()
+config.read(r".\configFile.ini")
+region_display = {"australiaeast": "Australia East",
+                  "eastasia": "East Asia", "eastus": "East US", "uksouth": "UK South"}
+
+
+def extract_and_update_results(test_run_response, test_run_id, location):
+    """
+    Download results Zip 
+    Extract the Zip file
+    Add a column to the results CSV file
+    Save the updated CSV file
+
+    Parameters
+    ----------
+    test_run_response : dict
+        test run results
+
+    Returns
+    -------
+    dict
+
+    Raises
+    ------
+    TypeError
+        If the test_run_response is not a dict 
+    """
+    if not isinstance(test_run_response, dict):
+        raise TypeError("test_run_response must be a dictionary")
+
+    results_zip_url = test_run_response['testArtifacts']['outputArtifacts']['resultFileInfo']['url']
+
+    csv_file_path = Path("./testRunResults")
+    engine_index = 1  # This sample test run uses only one engine. In case of multiple engines, iterate over the result CSV files from each engine
+
+    zipfile.ZipFile(io.BytesIO(requests.get(
+        results_zip_url).content)).extractall(csv_file_path)
+
+    results_df = pd.read_csv(
+        csv_file_path/f"engine{str(engine_index)}_results.csv")
+    results_df["testRunName"] = test_run_response["displayName"]
+    results_df["Region"] = region_display[location]
+
+    csv_file_name = f"csvTestResults_{test_run_id}_testengine{str(engine_index)}.csv"
+    results_df.to_csv(
+        csv_file_path/csv_file_name,
+        index=False,
+        header=True,
+    )
+    return {"csv_file_path": csv_file_path, "csv_file_name": csv_file_name}
+
+
+def create_resources(mgmtClient, regions):
+    """
+    Create a resource in each region 
+    Get the resources details
+    Append to a list
+
+    Parameters
+    ----------
+    mgmtClient : LoadTestMgmtClient,
+    regions: string    
+
+    Returns
+    -------
+    list
+
+    Raises
+    ------
+    HttpResponseError
+        If resource creation or getting resource details fails 
+    """
+
+    resource_group = config["DEFAULT"]["resource_group"]
+    loadtest_resource_name = config["DEFAULT"]["loadtest_resource_name"]
+    try:
+        for region in regions:
+            resourceCreationPoller = mgmtClient.load_tests.begin_create_or_update(

```

Comment:
This is sequential creation. We can parallelize by starting all operations async operations and wait for resource creation in the same flow. This way we can eliminate time.sleep

https://learn.microsoft.com/en-us/python/api/azure-mgmt-loadtesting/azure.mgmt.loadtesting.aio.operations.loadtestsoperations?view=azure-python

Reply:
Reference: https://stackoverflow.com/questions/51696235/multiple-tasks-using-python-asyncio